kaiax/staking/staking_info.go

```diff
@@ -0,0 +1,215 @@
+	rList := make([]common.Address, 0, len(si.RewardAddrs))
+
+	for i := range si.NodeIds {
+		r := si.RewardAddrs[i]
+		if cn, ok := cmap[r]; ok {
+			cn.NodeIds = append(cn.NodeIds, si.NodeIds[i])
+			cn.StakingContracts = append(cn.StakingContracts, si.StakingContracts[i])
+			cn.StakingAmount += si.StakingAmounts[i]
+		} else {
+			cmap[r] = &consolidatedNode{
+				NodeIds:          []common.Address{si.NodeIds[i]},
+				StakingContracts: []common.Address{si.StakingContracts[i]},
+				RewardAddr:       r,
+				StakingAmount:    si.StakingAmounts[i],
+			}
+			rList = append(rList, r)
+		}
+	}
+
+	carr := make([]consolidatedNode, 0, len(cmap))
+	for _, r := range rList {
+		carr = append(carr, *cmap[r])
+	}
+	return &carr
+}
+
+// Returns the Gini coefficient among the staking amounts that are greater than or equal to minStake.
+// The amounts are first consolidated by RewardAddr, filtered by minStake, and then summarized to Gini.
+func (si *StakingInfo) Gini(minStake uint64) float64 {
+	// Cache hits only if the same minStake is used.
+	if si.cachedGini == nil || si.cachedGiniMinStake != minStake {
+		g := si.computeGini(minStake)
+		si.cachedGini = &g
+		si.cachedGiniMinStake = minStake
+	}
+	return *si.cachedGini
+}
+
+func (si *StakingInfo) computeGini(minStake uint64) float64 {
+	cnodes := si.ConsolidatedNodes()
+	amounts := make(sort.Float64Slice, 0, len(cnodes))
+
+	for _, cnode := range cnodes {
+		if cnode.StakingAmount >= minStake {
+			amounts = append(amounts, float64(cnode.StakingAmount))
+		}
+	}
+
+	return computeGini(amounts)
+}
+
+func computeGini(amounts sort.Float64Slice) float64 {
+	if len(amounts) == 0 {
+		return EmptyGini
+	}
+
+	sort.Sort(amounts)
+
+	sumOfAbsoluteDifferences := float64(0)
+	subSum := float64(0)
+
+	for i, x := range amounts {
+		temp := x*float64(i) - subSum
+		sumOfAbsoluteDifferences = sumOfAbsoluteDifferences + temp
+		subSum = subSum + x
+	}
+
+	result := sumOfAbsoluteDifferences / subSum / float64(len(amounts))
+	result = math.Round(result*100) / 100
+	return result
+}
+
+// Parse from persisted data
+func (sl *StakingInfoLegacy) ToStakingInfo() *StakingInfo {
+	si := &sl.StakingInfo
+
+	// Try to fill treasury fund addresses from legacy fields that may have been persisted in the past.
+	if !common.EmptyAddress(sl.KCFAddr) {
+		si.KEFAddr = sl.KCFAddr
+	}
+	if !common.EmptyAddress(sl.KFFAddr) {
+		si.KIFAddr = sl.KFFAddr
+	}
+	if !common.EmptyAddress(sl.KIRAddr) {
+		si.KEFAddr = sl.KIRAddr
+	}
+	if !common.EmptyAddress(sl.PoCAddr) {
+		si.KIFAddr = sl.PoCAddr
+	}
+
+	return si
+}
+
+// Convert to API response
+func (si *StakingInfo) ToResponse(useGini bool, minStake uint64) *StakingInfoResponse {
+	return &StakingInfoResponse{
+		StakingInfoLegacy: StakingInfoLegacy{
+			StakingInfo: *si,
+			KIRAddr:     si.KEFAddr,
+			PoCAddr:     si.KIFAddr,
+			KCFAddr:     si.KEFAddr,
+			KFFAddr:     si.KIFAddr,
+		},
+		UseGini: useGini,
+		Gini:    si.Gini(minStake),
```

Comment:
Is it your intended display regardless of `UseGini`?

Reply:
Yes, I intended to display the Gini value even if it's irrelevant. It's just a computed number.